warehouse/api/schema.py

```diff
@@ -0,0 +1,195 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+
+from marshmallow import Schema, fields
+
+
+class File(Schema):
+    filename = fields.Str()
+    packagetype = fields.Str()
+    python_version = fields.Str()
+    has_sig = fields.Bool(attribute="has_signature")
+    comment_text = fields.Str()
+    md5_digest = fields.Str()
+    digests = fields.Method("get_digests")
+    size = fields.Int()
+    upload_time = fields.Function(
+        lambda obj: obj.upload_time.strftime("%Y-%m-%dT%H:%M:%S")
+    )
+    url = fields.Method("get_detail_url")
+
+    def get_digests(self, obj):
+        return {"md5": obj.md5_digest, "sha256": obj.sha256_digest}
+
+    def get_detail_url(self, obj):
+        request = self.context.get("request")
+        return request.route_url("packaging.file", path=obj.path)
+
+
+class Release(Schema):
+    bugtrack_url = fields.Str(attribute="project.bugtrack_url")
+    classifiers = fields.List(fields.Str())
+    docs_url = fields.Str(attribute="project.documentation_url")
+    downloads = fields.Method("get_downloads")
+    project_url = fields.Method("get_project_url")
+    url = fields.Method("get_release_url")
+    requires_dist = fields.List(fields.Str())
+    files_url = fields.Method("get_files_url")
+
+    def get_files_url(self, obj):
+        request = self.context.get("request")
+        return request.route_url(
+            "api.views.projects.releases.files",
+            name=obj.project.name,
+            version=obj.version,
+        )
+
+    def get_project_url(self, obj):
+        request = self.context.get("request")
+        return request.route_url("api.views.projects.detail", name=obj.project.name)
+
+    def get_release_url(self, obj):
+        request = self.context.get("request")
+        return request.route_url(
+            "api.views.projects.releases.detail",
+            name=obj.project.name,
+            version=obj.version,
+        )
+
+    def get_downloads(self, obj):
+        return {"last_day": -1, "last_week": -1, "last_month": -1}
+
+    class Meta:
+        fields = (
+            "author",
+            "author_email",
+            "bugtrack_url",
+            "classifiers",
+            "description",
+            "description_content_type",
+            "docs_url",
+            "downloads",
+            "download_url",
+            "home_page",
+            "keywords",
+            "license",
+            "maintainer",
+            "maintainer_email",
+            "name",
+            "project_url",
+            "url",
+            "platform",
+            "requires_dist",
+            "requires_python",
+            "summary",
+            "version",
+            "files_url",
+        )
+        ordered = True
+
+
+class Project(Schema):
+    url = fields.Method("get_detail_url")
+    releases_url = fields.Method("get_releases_url")
```

Comment:
Drive-by comment: Since serializing URLs is a common usage, you might want to create a custom field. It could look something like

```python
class Project(Schema):
    releases_url = RouteURL("api.views.projects.releases",
                            name="<normalized_name>")
```

For an example of how this might be implemented, check out flask-marshmallow's equivalent: https://github.com/marshmallow-code/flask-marshmallow/blob/7e594dc89ba853946ef1593e9ac19850a5ec9d2a/flask_marshmallow/fields.py#L39